client/client.go

```diff
@@ -109,71 +103,86 @@ func (c *LogClient) ListByIndex(ctx context.Context, start, count int64) ([]*tri
 	return resp.Leaves, nil
 }
 
-// waitForRootUpdate repeatedly fetches the Root until the TreeSize changes
-// or until ctx times out.
-func (c *LogClient) waitForRootUpdate(ctx context.Context) error {
+// WaitForRootUpdate repeatedly fetches the Root until the fetched tree size >=
+// waitForTreeSize or until ctx times out.
+func (c *LogClient) WaitForRootUpdate(ctx context.Context, waitForTreeSize int64) (*trillian.SignedLogRoot, error) {
 	b := &backoff.Backoff{
 		Min:    100 * time.Millisecond,
 		Max:    10 * time.Second,
 		Factor: 2,
 		Jitter: true,
 	}
-	startTreeSize := c.root.TreeSize
 	for i := 0; ; i++ {
-		if err := c.UpdateRoot(ctx); err != nil {
-			return err
-		}
-		if c.root.TreeSize > startTreeSize {
-			return nil
+		root, err := c.UpdateRoot(ctx)
+		switch x := status.Code(err); x {
+		case codes.OK:
+			if root.TreeSize >= waitForTreeSize {
+				return root, nil
+			}
+		case codes.Unavailable: // Retry.
+		default:
+			return nil, err
 		}
-		if err := ctx.Err(); err != nil {
-			return status.Errorf(codes.DeadlineExceeded,
-				"%v. TreeSize: %v, want > %v. Tried %v times.",
-				err, c.root.TreeSize, startTreeSize, i+1)
+
+		select {
+		case <-ctx.Done():
+			return nil, status.Errorf(codes.DeadlineExceeded,
+				"%v. TreeSize: %v, want >= %v. Tried %v times: %v",
+				err, c.root.TreeSize, waitForTreeSize, i+1, ctx.Err())
+		case <-time.After(b.Duration()):
 		}
-		time.Sleep(b.Duration())
 	}
 }
 
-// UpdateRoot retrieves the current SignedLogRoot.
-// Verifies the signature, and the consistency proof if this is not the first root this client has seen.
-func (c *LogClient) UpdateRoot(ctx context.Context) error {
+// GetLatestRoot fetches and verifies the latest root against a trusted root, seen in the past.
+// Pass nil for trusted if this is the first time querying this log.
+func (c *LogClient) GetLatestRoot(ctx context.Context, trusted *trillian.SignedLogRoot) (*trillian.SignedLogRoot, error) {
```

Comment:
The naming seems a little reversed here - it feels like:

```golang
func (c) GetLatestRoot(ctx) (*trillian.SignedLogRoot, error) {
  untrusted, err := fetchLatestAndVerifySigs()
  ...
  if err := c.UpdateRoot(untrusted); err != nil {
   return nil, err
  }
  return untrusted, nil
}
 
func (c) UpdateRoot(..., untrusted *trillian.SignedLogRoot) error {
   proof, err := fetchConsistencyProof()
   ...
   if err := verifyConsistencyProof(c.trusted, untrusted); err != nil {
     return err
   }
   if untrusted.timestamp > c.trusted.timestamp && untrusted.size >= c.trusted.size {
      c.trusted = untrusted
   }
   return nil
}

```

is a bit more natural?

That'd allow you to pass in your own set of untrusted roots to `UpdateRoot()` if you wanted too (I'm guessing that's why you've added the param?)

Reply:
Hmm.. 
`GetLatestRoot` fetches and compares a new root to a previously trusted root. This is why you pass in a "currently" trusted root or nil to validate the new root against. 

`UpdateRoot` fetches and verifies the `LogClient`'s internal trusted root, verifying consistency with the last root the `LogClient` has seen.  `NewLogClient` supports supplying an initial trusted root, but from there on, we want to maintain a single chain of consistency proofs between the roots that the `LogClient` trusts.

Reply:
Making `getLatestRoot` private sounds good to me.

Reply:
`Update` has been renamed to `UpdateRoot`